presto-main/src/main/java/io/prestosql/sql/analyzer/AggregationAnalyzer.java

```diff
@@ -467,12 +467,23 @@ protected Boolean visitIdentifier(Identifier node, Void context)
             if (analysis.getLambdaArgumentReferences().containsKey(NodeRef.of(node))) {
                 return true;
             }
+
+            if (!ScopeReferenceExtractor.hasReferencesToScope(node, analysis, sourceScope)) {
+                // reference to outer scope is group-invariant
+                return true;
+            }
+
             return isGroupingKey(node);
         }
 
         @Override
         protected Boolean visitDereferenceExpression(DereferenceExpression node, Void context)
         {
+            if (!ScopeReferenceExtractor.hasReferencesToScope(node, analysis, sourceScope)) {
```

Comment:
Are we sure that won't return false positive? Maybe `DereferenceExpression node` does not reference any source fields entirely, but some base part of it does.

Reply:
Think of source produces `row<a> b`, then `b.a` won't reference source scope, but `a` does.

Reply:
`hasReferencesToScope` returns true if any subexpression is a reference to the scope. 

> Think of source produces row<a> b, then b.a won't reference source scope, but a does.

Did you mean the other way around? If the source column is named `b`, `b.a` will reference the source scope via `b`. `a` won't, since it's not a valid column of the source, but then the query would fail due to missing field.

Reply:
> hasReferencesToScope returns true if any subexpression is a reference to the scope.

Indeed.

> since it's not a valid column of the source, but then the query would fail due to missing field.

Query might still not fail, because `b` could be row and `a` could be element in row.

Anyway, I think this code is OK. Could you add some test that validates it works in case we described, that is source column `b` of type row (with `a` element), and subquery referencing `b.a`?